strict-version-matcher-plugin/src/main/java/com/google/android/gms/dependencies/DependencyInspector.java

```diff
@@ -0,0 +1,288 @@
+    if (resolvableDependencies.getResolutionResult() != null &&
+        resolvableDependencies.getResolutionResult().getAllDependencies() != null) {
+      registerDependencies(resolvableDependencies, projectName, taskName);
+    }
+
+    // Phase 2: take the resolved versions of Artifacts, go get the dependencies that
+    // apply to those specific versions, and then ensure all are being honored.
+    logger.info("Starting dependency analysis");
+    ResolutionResult resolutionResult = resolvableDependencies.getResolutionResult();
+
+    // Create an Artifact to ArtifactVersion mapping for resolved components.
+    HashMap<Artifact, ArtifactVersion> resolvedVersions = new HashMap<>();
+    for (ResolvedComponentResult resolvedComponentResult :
+        resolutionResult.getAllComponents()) {
+      ArtifactVersion version = ArtifactVersion.Companion.fromGradleRefOrNull(
+          resolvedComponentResult.getId().toString());
+      if (version != null) {
+        resolvedVersions.put(version.getArtifact(), version);
+      }
+    }
+
+    // Quick no-op when no versions.
+    if (resolvedVersions.size() < 1) {
+      return;
+    }
+
+    // Retrieve dependencies that apply to the resolved dep set.
+    Collection<Dependency> activeDeps = dependencyAnalyzer.getActiveDependencies(
+        resolvedVersions.values());
+    // Validate each of the dependencies that should apply.
+    for (Dependency dep : activeDeps) {
+      ArtifactVersion resolvedVersion = resolvedVersions.get(dep.getToArtifact());
+
+      // Check whether dependency is still valid.
+      if (!dep.isVersionCompatible(resolvedVersion.getVersion())) {
+        // This means a resolved version failed a dependency rule.
+
+        logger.warn("Dependency resolved to an incompatible version: " + dep);
+
+        // TODO: Warn, not fail, when the Major version boundaries are breached.
+        // TODO: Experiment with collecting all issues and reporting them at once.
+        Collection<Node> depsPaths = dependencyAnalyzer.getPaths(
+            resolvedVersion.getArtifact());
+
+        // Print extended path information at INFO level.
+        logger.info("Dependency Resolution Help: Displaying all currently known " +
+            "paths to any version of the dependency: " + dep.getToArtifact());
+        logger.info("NOTE: com.google.android.gms translated to c.g.a.g for brevity. " +
+            "Same for com.google.firebase -> c.g.f");
+        // TODO: The depPaths nodes need to be consolidated prior to display.
+        for (Node n : depsPaths) {
+          printNode(1, n);
+        }
+
+        throw new GradleException(getErrorMessage(dep, resolvedVersion, depsPaths));
+      }
+    }
+  }
+
+  @NotNull
+  private String getErrorMessage(@Nonnull Dependency dep, @Nonnull ArtifactVersion resolvedVersion,
+                                 @Nonnull Collection<Node> depPaths) {
+    StringBuilder errorMessage = new StringBuilder("In project '" + projectName +
```

Comment:
if possible, break strings on newlines.

ex: "an error\n" + "involving newlines."

This makes this block a bit more readable as the error it will generate because format will treat it correctly

Alternatively, the pattern you have below with multiple .append calls to this stringbuilder is also acceptable.

Reply:
Done.